Billing.API/Controllers/InvoiceController.cs

```diff
@@ -51,12 +46,51 @@ public async Task<IActionResult> GetInvoices([FromRoute] string origin, [FromRou
             return Ok(list);
         }
 
+        [HttpGet("/accounts/{origin}/{clientId:int:min(0)}/invoice/{filename}")]
+        public async Task<IActionResult> GetInvoices([FromRoute] string origin, [FromRoute] int clientId, [FromRoute] string filename)
+        {
+            _logger.LogDebug("Getting invoice for {0} client {1} filename {2}", origin, clientId, filename);
+
+            string clientPrefix;
+
+            try
+            {
+                clientPrefix = GetClientPrefix(origin);
+            }
+            catch
+            {
+                return BadRequest();
+            }
+
+            var fileParts = filename.Split('_');
+
+            var fileId = fileParts[2].ToLowerInvariant().ReplaceByEmpty(".pdf").ToInt32();
```

Comment:
What do you think about using regular expressions to parse the name using a regular expression? In my opinion, it will be explicit and cleaner.

Reply:
Or... maybe you can put the pattern in the URL binding pattern?

Something like this (I did not tested it):

```
[HttpGet("/accounts/{origin}/{clientId:int:min(0)}/invoice/{prefix}_{fileId:int:min(1)}.pdf")]
```

Reply:
I used this regex, since the last part is what we need:

```csharp
const string pattern = @"(\d+)\.(\w+)$";
```

is it ok?

Reply:
I was thinking about something more restrictive and explicit, what do you think about this:

```suggestion
            private static readonly _invoiceFilenameRegex = new Regex(@"^invoice_\d\d\d\d-\d\d-\d\d_(\d+).pdf$");
            // . . .
            var fileIdStr = 
            var fileParts = filename.Split('_');
            var fileId = _invoiceFilenameRegex.Match(filename).Groups[1].Value.ToInt32();
```

Reply:
Excellent